### PR TITLE
Exercise 8

### PR DESCRIPTION
**Pull request by angeline-tech**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,13 +5,22 @@ before_install:
...
-- docker run -e FLASK_APP=$FLASK_APP -e FLASK_ENV=$FLASK_ENV -e TRELLO_API_KEY=$TRELLO_API_KEY -e TRELLO_TOKEN=$TRELLO_TOKEN todo_app:test acceptance
+- docker run -e FLASK_APP=$FLASK_APP -e FLASK_ENV=$FLASK_ENV -e TRELLO_API_KEY=$TRELLO_API_KEY
+  -e TRELLO_TOKEN=$TRELLO_TOKEN todo_app:test acceptance
+- docker build --target production --tag $DOCKER_USERNAME/todo_app:production .
+- echo "$DOCKER_TOKEN" | docker login -u "$DOCKER_USERNAME" --password-stdin
+- docker push $DOCKER_USERNAME/todo_app:production
+- docker login --username=$HEROKU_LOGIN --password=$HEROKU_API_KEY registry.heroku.com
+- docker tag $DOCKER_USERNAME/todo_app:production registry.heroku.com/$DOCKER_USERNAME-todo-app/web
+- docker push registry.heroku.com/$DOCKER_USERNAME-todo-app/web
+- heroku container:release web -a $DOCKER_USERNAME-todo-app
```
Comment (JackMead) [inline .travis.yml:16]: Currently all of these steps take place during your script phase - that means that you can't separate your testing verification from a deployment. Instead we could take advantage of the `before_deploy` ([mentioned in the docs here](https://docs.travis-ci.com/user/job-lifecycle/#deploying-your-code)) and/or `deploy` travis lifecycle stages (see the [script deployment](https://docs.travis-ci.com/user/deployment/script/)) - these won't run on PRs by default, and we can restrict which branches they run on

**Review by JackMead — CHANGES_REQUESTED** [review]
Nice work, with a little detective work I was able to track down what I _think_ is your Heroku app (https://angeline23-todo-app.herokuapp.com/) which is up and running happily. There are a couple of small tweaks that we should make; one is to document this in your Readme to confirm where the app is hosted and when the pipeline should run, and the other is as per the comment below - we should ensure that this deployment only happens on appropriate branch pushes - we don't want every PR to automatically deploy the environment before it's been reviewed.

**Comment by hugh-emerson** [comment]
The deploy section you've written doesn't work. 

Did you take a look at the link Jack shared?
https://docs.travis-ci.com/user/deployment/script/

It explains how to use it and even gives an example. Please contact us if you need help getting it to work.

You can see that the build failed after you merged:
https://app.travis-ci.com/github/angeline-tech/DevOps-Course-Starter/builds/231666292

Also, I think you uploaded this as a note rather than submitting the module. Please use the link on the right hand side of the unit in Aptem, and make sure to include a link to the pull request on GitHub. Otherwise it won't get tracked correctly

**Comment by angeline-tech** [comment]
@hugh-emerson I've fixed the deployment and raised a new PR since this one was closed.
https://github.com/angeline-tech/DevOps-Course-Starter/pull/11 
I didn't notice the deployment was broken because it only deploys from master and I pushed to exercise-8.
This build shows it working, I've set it back to deploying only from master now.
https://travis-ci.com/github/angeline-tech/DevOps-Course-Starter/builds/234131141

**Pull request merged**